### PR TITLE
Add a mechanism to lazily obtain realm token

### PR DESCRIPTION
**Pull request by FadhlanR**
Tickets: CS-6505 & CS-6511

After discussions with @ef4, I  introduced a new function, `fetchWithAuth`, within the `LoaderService`. This function serves as a wrapper around `loader.fetch`, enhancing it with additional functionalities:

1. **Authorization Header Inclusion:** Before initiating the `loader.fetch` operation, `fetchWithAuth` automatically includes the `Authorization` token in the request header.

2. **Unauthorized Response Handler:** Following the `loader.fetch` operation, `fetchWithAuth` incorporates a mechanism to handle unauthorized responses gracefully.

### CODE REVIEW

File: packages/host/app/services/sessions-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/sessions-service.ts]: Can this method figure out whether to skip cache by itself so we don't have to tell it to? It could check if the jwt is about to expire or did expire and skip the cache in that case

File: packages/host/app/services/sessions-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/sessions-service.ts]: We are doing a lot of the same things compared to [my PR](https://github.com/cardstack/boxel/pull/1014/files#diff-24d76ba72eb2961323a9fbeabd16ee1c9a1c195583e4b44ef94fb9d6aee3ea94R22), we should consolidate the logic, ideally reuse

**Review by lukemelia — COMMENTED** [review]
Are there cases that we have refactored here that could be changing which loader instance is in use? I'm not really sure where we are (and where we are going) with respect to multiple loaders, but I seem to recall @habdelra working to make that possible.

File: packages/host/app/services/sessions-service.ts
Comment (FadhlanR) [reply packages/host/app/services/sessions-service.ts]: There is also a case where the JWT is still in valid period but the realm owner update the user permissions. So, this `skipCache` parameter is used to handle that case.

File: packages/host/app/services/sessions-service.ts
Comment (FadhlanR) [reply packages/host/app/services/sessions-service.ts]: I think I can reuse your `RealmAuthClient` class.

**Comment by FadhlanR** [comment]
> Are there cases that we have refactored here that could be changing which loader instance is in use? I'm not really sure where we are (and where we are going) with respect to multiple loaders, but I seem to recall @habdelra working to make that possible.

Good point, I wasn't aware of this. There are a few resources that receive the loader instance as a parameter, but it appears we only use it in `CardResource`. Are there any other places where we pass the loader instance as a parameter?

File: packages/host/app/resources/card-type.ts
Comment (habdelra) [inline packages/host/app/resources/card-type.ts]: i would recommend trying to limit our use of the loader service if at all possible. instead of relying on the loader service, let's pass in the loader

File: packages/host/app/resources/card-type.ts
Comment (habdelra) [inline packages/host/app/resources/card-type.ts]: add a loader as part of the thunks used to create the resource

File: packages/host/app/resources/import.ts
Comment (habdelra) [inline packages/host/app/resources/import.ts]: same comment as above--let's try to limit our use of the loader service when possible

File: packages/host/app/services/card-service.ts
Comment (habdelra) [inline packages/host/app/services/card-service.ts]: same comment--lets try to limit our reliance on the loader service--if the caller passes in a loader, we should use it

**Comment by habdelra** [comment]
I'm seeing that you are undoing a lot of the work that @lukemelia mentioned around supporting multiple loaders--specially there are lots of places where you are removing the loader as a param and relying on the loader service. This moves us farther away from being able to support multiple loaders--and rather forces us to use a single loader. what's the rationale for this change?

**Comment by habdelra** [comment]
thinking about this more, I think that `fetchWithAuth` probably doesn't live in the loader service, since by doing that it forces us to use a single loader in the host app (the loader being used by the loader service). We have already identified that that is what we want to move away from--i.e. having a loader per stack. Rather, I think `fetchWithAuth()` should take as one of it's parameters the loader to use. Perhaps this can be a function that doesn't live in any service--but rather just an exported function in the lib/ dir of the host app like our other utils.

EDIT: 
also it does look like it needs access to the DI container since it requires the realm service to figure out what realm the requested resource is in. so perhaps that is another parameter: `owner`--from which you can lookup the realm service (using `getOwner().lookup('service:realm')`. I strongly question the need to use the session service--since the way it is written it's making fundamental misconceptions about realms. Probably until CS-6550 is fixed, it should be careful about using the session service.

**Comment by habdelra** [comment]
Note PR https://github.com/cardstack/boxel/pull/1026, this is what you'll want to use instead of the `SessionService`

**Comment by FadhlanR** [comment]
> Note PR #1026, this is what you'll want to use instead of the `SessionService`

Noted! Please hold off the review, I'll continue to work on this PR after @habdelra's PR gets merged.

**Comment by FadhlanR** [comment]
This PR is ready for another review. Based on our pairing session, I decided to make `fetchWithAuth` function as `URLHandler` and register it to the `Loader` using `prependURLHandlers`. So I can avoid to remove the codes related to multiple loaders although we have plan to not have multiple loaders.

**Comment by jurgenwerk** [comment]
Can you explain how to test this? 

Locally I set my drafts realm to be readable only by a specific user and then I wanted to see how the host would obtain a JWT  to access things from drafts realm. But I could not get past this:

image

It looks like where this is making a HEAD request to the draft's realm index there is no `x-boxel-realm-url` header present which stops the feature from working

File: packages/host/app/resources/realm-session.ts
Comment (jurgenwerk) [inline packages/host/app/resources/realm-session.ts]: Perhaps `rawRealmToken` is a better name for this getter. It tells you straight away that this is a raw string to be included in the header.

File: packages/host/app/resources/realm-session.ts
Comment (jurgenwerk) [inline packages/host/app/resources/realm-session.ts]: Can this be private?

File: packages/host/app/services/loader-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/loader-service.ts:22]: Let's add a note on why exactly it's safe to keep the cache here - if I'm not mistaken it's because the owner is the service, which stays around for the whole lifetime of the host app, which in turn assures the resources will not get torn down?

File: packages/host/app/services/loader-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/loader-service.ts]: Perhaps `realmSessions` is a better name for this map?

File: packages/host/app/services/loader-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/loader-service.ts]: `request.url.includes('session')` will include any url with the word session in it - is that ok or did you mean the specific form of the realm's session endpoint?

File: packages/host/app/services/loader-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/loader-service.ts]: Would it be better to rename this to `getRealmSession` or `getRealmSessionResource`?

File: packages/host/app/services/matrix-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/matrix-service.ts]: Why change to warn? Should this not be a hard error?

File: packages/host/app/services/realm-info-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/realm-info-service.ts]: Should this be an error instead? I mean if there is no realm header then something is clearly wrong?

File: packages/host/app/services/realm-info-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/realm-info-service.ts:24]: How come there is a separate map for this? Would it be more organized to include this info in `RealmInfo`?

File: packages/host/app/services/realm-info-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/realm-info-service.ts]: For efficiency put booleans first when using && or ||

Comment (jurgenwerk) [reply packages/host/app/services/realm-info-service.ts]: Actually I don't see `skipCache` is actually being used where `this.realmInfoService.isPublicReadable()` is called

File: packages/host/app/services/realm-info-service.ts
Comment (jurgenwerk) [inline packages/host/app/services/realm-info-service.ts]: Is it ok to expect realmInfo as undefined? Don't we want this to always be present when we give it a URL? And throw an error if not authorized to get this info

File: packages/host/app/services/loader-service.ts
Comment (FadhlanR) [reply packages/host/app/services/loader-service.ts]: My idea is we don't need to include `Authorization` header, if the request is an authentication request. The way to differentiate it is by checking the `url`.

**Comment by FadhlanR** [comment]
> Can you explain how to test this?
> 
> Locally I set my drafts realm to be readable only by a specific user and then I wanted to see how the host would obtain a JWT to access things from drafts realm. But I could not get past this:
> 
> image
> It looks like where this is making a HEAD request to the draft's realm index there is no `x-boxel-realm-url` header present which stops the feature from working

We haven't include JWT on indexing so I don't think we can restrict read permission. Instead you can set permission like this:

`"http://localhost:4201/drafts/": {
    "users": {
      "*": ["read"],
      "@matic:localhost": ["read", "write"]
    }
  }`
  
So, only your account has permission to write to that realm. And you can try to edit any card instance from interact/code mode. After that, you can check the PATCH request header.

Screenshot 2024-02-15 at 12 19 45

File: packages/host/app/services/realm-info-service.ts
Comment (FadhlanR) [reply packages/host/app/services/realm-info-service.ts]: I made this function more general by renaming the parameter to `url`. This change implies that clients can pass any URLs, not necessarily URLs that are part of the realm. Therefore, the result can be undefined if the URL is not associated with a realm. Do you think it is better to remove the `console.war`?

File: packages/host/app/services/loader-service.ts
Comment (FadhlanR) [reply packages/host/app/services/loader-service.ts]: And I have to update the realm checkPermission. [bcba1c2](https://github.com/cardstack/boxel/pull/1020/commits/bcba1c2f2a3979c1c36948bf2fa56b4678da1365)

File: packages/host/app/services/matrix-service.ts
Comment (FadhlanR) [reply packages/host/app/services/matrix-service.ts]: reverted it here [060e44d](https://github.com/cardstack/boxel/pull/1020/commits/060e44db588f70f88c23ba4e728012099335f55f).

File: packages/host/app/services/realm-info-service.ts
Comment (FadhlanR) [reply packages/host/app/services/realm-info-service.ts]: Please check this commit [458ac6d](https://github.com/cardstack/boxel/pull/1020/commits/458ac6dec87659c50e325d90a9b3618c77a8740f).

File: packages/host/app/services/realm-info-service.ts
Comment (FadhlanR) [reply packages/host/app/services/realm-info-service.ts]: I updated it here [458ac6d](https://github.com/cardstack/boxel/pull/1020/commits/458ac6dec87659c50e325d90a9b3618c77a8740f).

**Comment by jurgenwerk** [comment]
@FadhlanR it is important we also test permissioned reading, not only writing. Even though the indexing doesn't support that yet, I believe we can still locally set up some permissioned realms (for example drafts) in a way that have restricted reading.

Here's my idea of how to test it:

1. You add your user to be able to read the drafts realm and nobody else:

```
{
  "http://localhost:4201/base/": {
    "users": {
      "*": ["read", "write"]
    }
  },
  "http://localhost:4201/drafts/": {
    "users": {
      "@user:localhost": ["read"]
    }
  },
  "http://localhost:4201/published/": {
    "users": {
      "*": ["read", "write"]
    }
  },
  "https://cardstack.com/base/": {
    "users": {
      "*": ["read", "write"]
    }
  }
}
```

2. Then, start the realms. This will produce problems when running `start:all` because the script will try to `http-get` a drafts realm as a health check but it cannot because that realm is now restricted to a single user and requires a JWT. The workaround is to temporarily remove that check. The script then becomes:


```
#! /bin/sh

NODE_NO_WARNINGS=1 start-server-and-test \
  'run-p start:matrix start:smtp start:development start:base:root' \
  'http-get://localhost:4201/base/fields/boolean-field?acceptHeader=application%2Fvnd.card%2Bjson|http-get://localhost:4203/fields/boolean-field?acceptHeader=application%2Fvnd.card%2Bjson|http-get://http://localhost:8008|http://localhost:5001' \
  'run-p start:test-realms start:test-container' \
  'http-get://localhost:4202/node-test/person-1?acceptHeader=application%2Fvnd.card%2Bjson|http-get://127.0.0.1:4205' \
  'wait'
```

At this point `start:all` should start the realms and you can see from the logs all the realms started and finished indexing without errors. 

I believe at this point a user `@user:localhost` should able to authenticate with the mechanism introduced in this PR and be able to read the drafts realm from within the host app. But there are errors that I pasted in my previous message.

File: packages/host/app/services/realm-info-service.ts
Comment (FadhlanR) [reply packages/host/app/services/realm-info-service.ts:24]: ~~I updated it to one map [41b7ef0](https://github.com/cardstack/boxel/pull/1020/commits/41b7ef0ed585ef057e07fcb6254ea914749d3028).~~ (EDIT) 

I just noticed we need different map for `RealmInfo` and `isPublicReadable`. For `RealmInfo`, we need `TrackedMap` for `RealmDropdown` component. So I reverted the changes.

**Comment by FadhlanR** [comment]
> @FadhlanR it is important we also test permissioned reading, not only writing. Even though the indexing doesn't support that yet, I believe we can still locally set up some permissioned realms (for example drafts) in a way that have restricted reading.
> 
> Here's my idea of how to test it:
> 
> 1. You add your user to be able to read the drafts realm and nobody else:
> 
> ```
> {
>   "http://localhost:4201/base/": {
>     "users": {
>       "*": ["read", "write"]
>     }
>   },
>   "http://localhost:4201/drafts/": {
>     "users": {
>       "@user:localhost": ["read"]
>     }
>   },
>   "http://localhost:4201/published/": {
>     "users": {
>       "*": ["read", "write"]
>     }
>   },
>   "https://cardstack.com/base/": {
>     "users": {
>       "*": ["read", "write"]
>     }
>   }
> }
> ```
> 
> 2. Then, start the realms. This will produce problems when running `start:all` because the script will try to `http-get` a drafts realm as a health check but it cannot because that realm is now restricted to a single user and requires a JWT. The workaround is to temporarily remove that check. The script then becomes:
> 
> ```
> #! /bin/sh
> 
> NODE_NO_WARNINGS=1 start-server-and-test \
>   'run-p start:matrix start:smtp start:development start:base:root' \
>   'http-get://localhost:4201/base/fields/boolean-field?acceptHeader=application%2Fvnd.card%2Bjson|http-get://localhost:4203/fields/boolean-field?acceptHeader=application%2Fvnd.card%2Bjson|http-get://http://localhost:8008|http://localhost:5001' \
>   'run-p start:test-realms start:test-container' \
>   'http-get://localhost:4202/node-test/person-1?acceptHeader=application%2Fvnd.card%2Bjson|http-get://127.0.0.1:4205' \
>   'wait'
> ```
> 
> At this point `start:all` should start the realms and you can see from the logs all the realms started and finished indexing without errors.
> 
> I believe at this point a user `@user:localhost` should able to authenticate with the mechanism introduced in this PR and be able to read the drafts realm from within the host app. But there are errors that I pasted in my previous message.

Based on discussion in boxel meeting, I think we can mark this scenario as out-of-scope.

File: packages/host/app/services/matrix-service.ts
Comment (habdelra) [inline packages/host/app/services/matrix-service.ts:245]: :tada:

**Review by habdelra — APPROVED** [review]
looks good. there was an upstream change to add `realmResource` property to the `FileResource`. Probably this PR should also rename that property to `realmSessionResource`.

File: packages/runtime-common/realm.ts
Comment (jurgenwerk) [inline packages/runtime-common/realm.ts]: `endsWith('_session')`?

**Review by jurgenwerk — APPROVED** [review]

**Pull request merged**
